lib/MemoryModel/MemModel.cpp

```diff
@@ -189,7 +189,7 @@ bool SymbolTableInfo::computeGepOffset(const llvm::User *V, LocationSet& ls) {
         // Handling array types, skipe array handling here
         // We treat whole array as one, then we can distinguish different field of an array of struct
         // e.g. s[1].f1 is differet from s[0].f2
-        if(isa<ArrayType>(*gi))
+        if(isa<ArrayType>(gi.getIndexedType()))
```

Comment:
Couldn't sleep, so poking at this a bit.

Yes, in LLVM 4 the gep_type_iterator no longer exposes the "outer" type,
[see here for a bit of explanation](https://github.com/llvm-mirror/llvm/blob/release_40/include/llvm/IR/GetElementPtrTypeIterator.h#L90).

I've only looked at MemModel.cpp (haven't tried the LocMemModel, that looks more difficult) but at least as far as the "fi_tests" from PTAben are concerned the following change seems to "fix" things:

```diff
diff --git a/lib/MemoryModel/MemModel.cpp b/lib/MemoryModel/MemModel.cpp
index 9bd65f9..01c0d17 100644
--- a/lib/MemoryModel/MemModel.cpp
+++ b/lib/MemoryModel/MemModel.cpp
@@ -193,7 +193,7 @@ bool SymbolTableInfo::computeGepOffset(const llvm::User *V, LocationSet& ls) {
         // Handling array types, skipe array handling here
         // We treat whole array as one, then we can distinguish different field of an array of struct
         // e.g. s[1].f1 is differet from s[0].f2
-        if(isa<ArrayType>(gi.getIndexedType()))
+        if(gi.isBoundedSequential())
             continue;
```
Thought I'd share, haven't worked on this enough to endorse this but might help regardless :).
For example, using `isSequential()` instead of `isBoundedSequential()` did *not* work properly on the tests, so more investigation is required I think.

Basically looking at the changes made in the commit that changed `gep_type_iterator`s behavior show how the various LLVM components were updated to accommodate this change (including BasicAA).

BridgedGEPIterator looks great, haven't had a chance to look at it yet, was trying to understand the change in behavior first...